File: golang/quick.go
package main

import "fmt"

// 第一种写法
func quickSort(values []int, left, right int) {
	temp := values[left]
	p := left
	i, j := left, right

	for i <= j {
		for j >= p && values[j] >= temp {
			j--
		}
		if j >= p {
			values[p] = values[j]
			p = j
		}

		if i <= p && values[i] <= temp {
			i++
		}
		if i <= p {
			values[p] = values[i]
			p = i
		}
	}

	values[p] = temp
	fmt.Println(values)

	if p-left > 1 {
		fmt.Println("sort left")
		quickSort(values, left, p-1)
	}
	if right-p > 1 {
		fmt.Println("sort right")
		quickSort(values, p+1, right)
	}
}

func QuickSort(values []int) {
	if len(values) <= 1 {
		return
	}
	quickSort(values, 0, len(values)-1)
}

// 第二种写法
func Quick2Sort(values []int) {
	if len(values) <= 1 {
		return
	}
	mid, i := values[0], 1
	head, tail := 0, len(values)-1
	for head < tail {
		fmt.Println(values)
		if values[i] > mid {
			values[i], values[tail] = values[tail], values[i]
			tail--
		} else {
			values[i], values[head] = values[head], values[i]
			head++
			i++
		}
	}
	values[head] = mid
	Quick2Sort(values[:head])
	Quick2Sort(values[head+1:])
}

func main() {
	a := []int{1, 9, 2, 4, 3, 5, 7, 8, 6, 0}
	fmt.Println("a is ", a)
	QuickSort(a)

	b := []int{1, 9, 2, 4, 3, 5, 7, 8, 6, 0}
	fmt.Println("b is ", b)
	Quick2Sort(b)
	fmt.Println("b is ", b)
}

File: java/CodeLife/src/sort/Quick.java
package sort;

/**
* 快速排序
* 平均时间复杂度 O(nlogn)
*/
public class Quick {

	private static int[] a = new int[]{5,7,8,9,6,2,4,3,1,5,6,4,3,2,1,7,8,9};

	public Quick(){

	}

	private void quicksort(int left,int right){
		if(left>=right)
			return;
		System.out.println("left="+left+" right="+right);

		int temp=a[left],i=left,j=right,t;
		while (i!=j) {
			while (a[j]>=temp && i<j) {
				j--;
			}
			while (a[i]<=temp && i<j) {
				i++;
			}
			if(i<j){
				t = a[i];
				a[i] = a[j];
				a[j] = t;
			}
		}
		//交换最左和指针最终停留位置的值
		if(left!=i){
			a[left]=a[i];
			a[i]=temp;
		}

		for(int s : a){
			System.out.print(s+" ");
		}
		System.out.println();

		this.quicksort(left, i-1);
		this.quicksort(i+1, right);
	}

	public static void main(String[] args) {
		Quick quick = new Quick();
		quick.quicksort(0, a.length-1);

		System.out.println("result:");
		for(int s : a){
			System.out.print(s+" ");
		}
	}
}
